examples/Example13/electrons.cu

```diff
@@ -0,0 +1,341 @@
+  fieldPropagatorConstBz fieldPropagatorBz(BzFieldValue);
+
+  int activeSize = active->size();
+  for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < activeSize; i += blockDim.x * gridDim.x) {
+    const int slot      = (*active)[i];
+    Track &currentTrack = electrons[slot];
+    auto volume         = currentTrack.navState.Top();
+    int volumeID        = volume->id();
+    // the MCC vector is indexed by the logical volume id
+    int lvolID     = volume->GetLogicalVolume()->id();
+    int theMCIndex = MCIndex[lvolID];
+
+    // Init a track with the needed data to call into G4HepEm.
+    G4HepEmElectronTrack elTrack;
+    G4HepEmTrack *theTrack = elTrack.GetTrack();
+    theTrack->SetEKin(currentTrack.energy);
+    theTrack->SetMCIndex(theMCIndex);
+    theTrack->SetOnBoundary(currentTrack.navState.IsOnBoundary());
+    theTrack->SetCharge(Charge);
+    G4HepEmMSCTrackData *mscData = elTrack.GetMSCTrackData();
+    mscData->fIsFirstStep        = currentTrack.initialRange < 0;
+    mscData->fInitialRange       = currentTrack.initialRange;
+
+    // Compute safety, needed for MSC step limit.
+    double safety = 0;
+    if (!currentTrack.navState.IsOnBoundary()) {
+      safety = BVHNavigator::ComputeSafety(currentTrack.pos, currentTrack.navState);
+    }
+    theTrack->SetSafety(safety);
+
+    RanluxppDoubleEngine rnge(&currentTrack.rngState);
+
+    // Sample the `number-of-interaction-left` and put it into the track.
+    for (int ip = 0; ip < 3; ++ip) {
+      double numIALeft = currentTrack.numIALeft[ip];
+      if (numIALeft <= 0) {
+        numIALeft                  = -std::log(currentTrack.Uniform());
+        currentTrack.numIALeft[ip] = numIALeft;
+      }
+      theTrack->SetNumIALeft(numIALeft, ip);
+    }
+
+    // Call G4HepEm to compute the physics step limit.
+    G4HepEmElectronManager::HowFar(&g4HepEmData, &g4HepEmPars, &elTrack, &rnge);
+
+    // Remember value of initial range for the next step.
+    currentTrack.initialRange = mscData->fInitialRange;
+
+    // Get result into variables.
+    double geometricalStepLengthFromPhysics = theTrack->GetGStepLength();
+    // The phyiscal step length is the amount that the particle experiences
+    // which might be longer than the geometrical step length due to MSC. As
+    // long as we call PerformContinuous in the same kernel we don't need to
+    // care, but we need to make this available when splitting the operations.
+    // double physicalStepLength = elTrack.GetPStepLength();
+    int winnerProcessIndex = theTrack->GetWinnerProcessIndex();
+    // Leave the range and MFP inside the G4HepEmTrack. If we split kernels, we
+    // also need to carry them over!
+
+    // Check if there's a volume boundary in between.
+    double geometryStepLength;
+    vecgeom::NavStateIndex nextState;
+    if (BzFieldValue != 0) {
+      geometryStepLength = fieldPropagatorBz.ComputeStepAndPropagatedState</*Relocate=*/false, BVHNavigator>(
```

Comment:
No, because the particle needs to travel across the curved path. There may be some optimizations where we can avoid querying the geometry if within the safety, but that wasn't even worth it with the `LoopNavigator` so I doubt it will be with the `BVHNavigator`

Reply:
I don't get the argument with the curved path, which would allow even longer geometry steps to stay within safety. I mean if the physics step is smaller than safety, there is no way that a boundary crossing may happen in this range, magnetic field or not. So why compute the distance?  And I would not extrapolate based on some old observation with LoopNavigator and a single test case. If physics is more dense, I would expect many physics steps in the same volume not needing to know the precise distance to boundary. So I would definitely try it, I don't agree it is a minor optimization.